### ISSUE TITLE
[QUESTION] Why torch.cuda.synchronize() is added after  the computational CUDA stream waits the P2P communication

### ISSUE THREAD

**Issue by sneaxiy** [NVIDIA/Megatron-LM#420]
Quite curious about the following codes. In L94, the CUDA stream used for OP computation would wait the CUDA stream used for P2P communication, and there would be no data race in different CUDA streams any more. In this way, it seems that L98 is useless, but why is this synchronization added? If we remove L98, is it possible that the training process would hang?

https://github.com/NVIDIA/Megatron-LM/blob/040eac9414ccbd1301ae16369c3044c5632b7e14/megatron/core/pipeline_parallel/p2p_communication.py#L91-L98

**Comment by deepakn94** [comment]
I think this was added because of a bug in `batch_isend_irecv()`; I don't exactly remember the nature of the failure (it could have been hangs, but it could have also been wrong losses). This may no longer be a problem with the latest PyTorch containers; we just haven't checked.

**Comment by Edenzzzz** [comment]
@deepakn94 I also wonder whether we should call wait() here instead of calling, say for backward_recv.wait() before the backward step in send_forward_recv_backward during warmup? Here seems that there's no overlap at all, as wait() blocks the compute stream
